### PR TITLE
opendkim pe23pdkupdate

### PR DESCRIPTION
**Pull request by Jeyselvaraj**
### Added

* New test case files config,install,opendkim,service, user spec has been developed and added . 
* Workflow file test.yaml has been added as part of the pe2023 update.
* Added back the FreeBSD, ubuntu ,debian os family into metadata.json

### Changed

* Updated pdk version3.2.0.1 to support Puppet v8
* Linting fixes
* linting missing datatype warnings
* Update class documentation
* Updated .gitignore to exclude .cache .pdk and .vscode
* updated the metadata dependencies into metadata
* anchors has been changed from the init.pp file into the contain function to overcome the warning from the validations but kept arrows to make sure the correct order to be followed.
* OS facts has been change from debian to CentOS and RedHat.
* opendkim spec file modified with the os based test cases .

### Removed
* .gemfile has been removed as got an validation error that should remove the either one of them as got another gemfile.

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -1,9 +1,41 @@
...
+# @param $manage_private_keys private keys of trusted host
+# @param $alldomain was the domain to be used when generating the signature
+# @param $selector was the selector to be used when generating the signature
+# @param $publickey The public key DNS record should appear as a TXT resource record
+# @param $publickeyextended The public key DNS record should appear as a TXT resource record
+# @param $privatekey  Generating a private and public key pair for use when signing your mail
+# @param $hash_algorithms encrypt algorithm
+# @param $signaturealgorithm signature of the algorithm
+# @param $minimumkeybits restrition of the minimum bits required
+# @api private
```
Comment (Ubeek) [inline manifests/init.pp]: Remove api private as this is the entrypoint for the module

File: manifests/init.pp [hunk 2]
```diff
@@ -38,25 +70,21 @@
...
-
-  anchor { 'opendkim::begin': }
-  -> class { '::opendkim::user': }
-  -> class { '::opendkim::install': }
-  -> class { '::opendkim::config': }
-  ~> class { '::opendkim::service': }
-  -> anchor { 'opendkim::end': }
-
+)  inherits opendkim::params {
+  contain 'opendkim::user'
```
Comment (Ubeek) [inline manifests/init.pp]: To be investigated, keep chaining/notification or drop back to contains?

File: manifests/params.pp [hunk 3]
```diff
@@ -35,24 +38,16 @@
   $service_enable    = true
   $service_ensure    = 'running'
 
-  case fact('os.family') {
```
Comment (Ubeek) [inline manifests/params.pp:38]: Revert back to including all OS's

File: manifests/params.pp [hunk 4]
```diff
@@ -35,24 +38,16 @@
...
-    'Debian': {
+  case $facts['os']['name'] {
```
Comment (Ubeek) [inline manifests/params.pp]: Revert back to testing 'family' not 'name'

File: pdk-release-1.20.0.deb
Comment (Ubeek) [inline pdk-release-1.20.0.deb:1]: Remove this file

File: pdk3.2.0-initial-validate-01.sh
Comment (Ubeek) [inline pdk3.2.0-initial-validate-01.sh:1]: Remove this file

File: spec/classes/config_spec.rb
Comment (Ubeek) [inline spec/classes/config_spec.rb:1]: Move all tests to opendkim_spec and remove this file

File: spec/classes/install_spec.rb
Comment (Ubeek) [inline spec/classes/install_spec.rb:1]: Remove

File: spec/classes/opendkim_spec.rb
Comment (Ubeek) [inline spec/classes/opendkim_spec.rb:1]: Move other tests into here

File: spec/classes/opendkim_spec.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+require 'spec_helper'
+
+describe 'opendkim' do
+  on_supported_os.each do |os, _facts|
+    context "on #{os}" do
+      let(:facts) { { 'os' => { 'name' => 'RedHat' } } }
+      let(:pre_condition) do
```
Comment (Ubeek) [inline spec/classes/opendkim_spec.rb]: remove precondition

File: spec/classes/opendkim_spec.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+require 'spec_helper'
+
+describe 'opendkim' do
+  on_supported_os.each do |os, _facts|
+    context "on #{os}" do
+      let(:facts) { { 'os' => { 'name' => 'RedHat' } } }
```
Comment (Ubeek) [inline spec/classes/opendkim_spec.rb]:
```suggestion
    let(:facts) do
      facts
    end
```

File: spec/classes/service_spec.rb
Comment (Ubeek) [inline spec/classes/service_spec.rb:1]: remove

File: spec/classes/user_spec.rb
Comment (Ubeek) [inline spec/classes/user_spec.rb:1]: Remove

**Review by Ubeek — CHANGES_REQUESTED** [review]

File: manifests/init.pp (hunk 1 above)
Comment (Jeyselvaraj) [reply manifests/init.pp]: removed the api private

File: pdk-release-1.20.0.deb
Comment (Jeyselvaraj) [reply pdk-release-1.20.0.deb:1]: deleted this file

File: pdk3.2.0-initial-validate-01.sh
Comment (Jeyselvaraj) [reply pdk3.2.0-initial-validate-01.sh:1]: removed

File: spec/classes/opendkim_spec.rb
Comment (Jeyselvaraj) [reply spec/classes/opendkim_spec.rb:1]: moved other tests into the opendkim_spec file

File: spec/classes/install_spec.rb
Comment (Jeyselvaraj) [reply spec/classes/install_spec.rb:1]: Removed done

File: spec/classes/user_spec.rb
Comment (Jeyselvaraj) [reply spec/classes/user_spec.rb:1]: Removed this file

File: spec/classes/opendkim_spec.rb (hunk 6 above)
Comment (Jeyselvaraj) [reply spec/classes/opendkim_spec.rb]: Thanks @Ubeek ,for this suggestion it did worked in the facts collection with the Rhel and Centos

File: spec/classes/opendkim_spec.rb (hunk 5 above)
Comment (Jeyselvaraj) [reply spec/classes/opendkim_spec.rb]: Removed the pre-conditions

File: manifests/params.pp (hunk 4 above)
Comment (Jeyselvaraj) [reply manifests/params.pp]: reverted to family

File: manifests/params.pp (hunk 3 above)
Comment (Jeyselvaraj) [reply manifests/params.pp:38]: included all OS

File: spec/classes/config_spec.rb
Comment (Jeyselvaraj) [reply spec/classes/config_spec.rb:1]: consolidated test cases into opendkim spec

File: spec/classes/service_spec.rb
Comment (Jeyselvaraj) [reply spec/classes/service_spec.rb:1]: removed the spec file

File: manifests/init.pp (hunk 2 above)
Comment (Jeyselvaraj) [reply manifests/init.pp]: Updated the code to contain the order with out order and kept order as puppet-lint considers them a potential issue for code readability.

File: CHANGELOG.md [hunk 7]
```diff
@@ -19,11 +20,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Update class documentation
 * Updated .gitignore to exclude .cache .pdk and .vscode
 * updated the metadata dependencies into metadata
-* anchors has been changed from the init.pp file into the contain function to overcome the warning from the validations.
+* anchors has been changed from the init.pp file into the contain function to overcome the warning from the validations but kept arrows to make sure the correct order to be followed.
 * OS facts has been change from debian to CentOS and RedHat.
+* opendkim psec file modified with the os based test cases .
```
Comment (Ubeek) [inline CHANGELOG.md]: Typo

File: manifests/init.pp [hunk 8]
```diff
@@ -87,4 +124,9 @@
   contain 'opendkim::install'
```
Comment (Ubeek) [inline manifests/init.pp]: Remove contains

**Review by Ubeek — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 7 above)
Comment (Jeyselvaraj) [reply CHANGELOG.md]: fixed

File: manifests/init.pp (hunk 8 above)
Comment (Jeyselvaraj) [reply manifests/init.pp]: Able to remove contain statement and now class included by absolute name. Now only operators ( -> and ~>) kept in the code

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -1,3 +1,75 @@
...
+  end
+end
+
+# read default_facts and merge them over what is provided by facterdb
+default_facts.each do |fact, value|
+  add_custom_fact fact, value, merge_facts: true
+end
+
+RSpec.configure do |c|
+  c.default_facts = default_facts
```
Comment (thallam-afg) [inline spec/spec_helper.rb:41]: Need to include hiera data

**Review by thallam-afg — APPROVED** [review]

**Pull request merged**